Deprecate TimeUnit2

batmat opened the pull request:
`java.util.concurrent.TimeUnit` is preferrable now.

Quoting Stephen:
> Java 5 did not have all the units required. So TU2 was one that had better conversion. Can be deprecated 
once we upgrade to Java 6 ;-)

# Description

* I didn't file a JENKINS issue since this is a pretty internal/dev oriented change. But if this is judged necessary, please tell me so and I'll do it.

Details: see title, very small change. And 0 risk from a runtime perspective. 

<!-- Comment: 
If the issue is not fully described in the ticket, please put additional comments (justification, pull request links, etc.).
-->

### Changelog entries

Proposed changelog entries:

_N/A_ I don't even think this should have an entry in the changelog, should it?

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- ~~[ ] JIRA issue is well described~~
- ~~[ ] Link to JIRA ticket in description, if appropriate~~
- [X] Appropriate autotests or explanation to why this change has no tests
- ~~[ ] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)~~

<!-- Comment: 
 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Desired reviewers

@reviewbybees

<!-- Comment:
If you want to get reviews from particular people, please CC them.
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

batmat committed:
Deprecate TimeUnit2

java.util.concurrent.TimeUnit is preferrable now.

Quoting Stephen:
"Java 5 did not have all the units required. So TU2 was
one that had better conversion. Can be deprecated once
we upgrade to Java 6 ;-)"

oleg-nenashev reviewed (approved):
:bee:

oleg-nenashev commented:
Maybe `@Restricted(DoNotUse.class)` will be a good addition

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

batmat commented:
@stephenc @oleg-nenashev `@Restricted(DoNotUse.class)` also added as requested/proposed.

batmat commented:
@reviewbybees done

batmat committed:
Also restrict its usage

jglick reviewed (changes requested):
There are dozens of usages of `TimeUnit2` in this repository. You should clean them up. (I know they are harmless, but as a matter of policy we should not be deprecating things with a claim that a full replacement is available, without proof that this is actually so in some representative code base.)

batmat commented:
Agreed @jglick I actually have that change locally, but it fails for now because accmod either is limited or, more probably IIUC, has a bug wrt to enums.

```
[INFO] --- access-modifier-checker:1.11:enforce (default) @ jenkins-core ---
[ERROR] hudson/util/TimeUnit2$1:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$2:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$7:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$4:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$3:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$6:0 hudson/util/TimeUnit2 must not be used
[ERROR] hudson/util/TimeUnit2$5:0 hudson/util/TimeUnit2 must not be used
```

jglick commented:
Arguably a bug in `access-modifier-checker`. You could fix that and `deploy` a snapshot for use in this PR, or just relax to `NoExternalUse` for now.

batmat committed:
Fix issue with accmod on self referencing

batmat commented:
@jglick fixed, build is passing on my machine. 
It's for now using a SNAPSHOT of https://github.com/kohsuke/access-modifier/issues/6 while it's not merged/released.

jglick commented on core/src/main/java/hudson/util/TimeUnit2.java:
```diff
@@ -63,7 +67,12 @@
  */
+@Deprecated
+@RestrictedSince("TODO")
+@Restricted(DoNotUse.class)
```
Do we actually need this to be `@Restricted`? Why? `@Deprecated` accomplishes everything we need, I think.

batmat replied:
It was recommended by @oleg-nenashev and encouraged by @stephenc I guess this avoids we possibly reintroduce some usages as we go @jglick ?

jglick replied:
I generally reserve restriction for newly introduced (or modified) APIs, or cases where ongoing use is clearly going to break, so this seems like overkill. I am not objecting, just FYI.

jglick commented on pom.xml:
```diff
@@ -95,7 +95,7 @@ THE SOFTWARE.
     <findbugs.failOnError>true</findbugs.failOnError>
     <test-annotations.version>1.2</test-annotations.version>
-    <access-modifier.version>1.11</access-modifier.version>
+    <access-modifier.version>1.12-PR6-SNAPSHOT</access-modifier.version>
```
You need to change the two properties below instead of this one.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

batmat committed:
Downgrade restriction from DoNotUse to NoExternalUse

There's a failure with accmod using DoNotUse. So taking the
easy path here to move forward. This will already have the intended
effect: clarify that TimeUnit2 is not to be used anymore.

batmat commented:
[Finally followed the recommendation of @jglick above](https://github.com/jenkinsci/jenkins/pull/2892#issuecomment-302710465):

> [...] just relax to NoExternalUse for now.

In the end, the goal of actually deprecating `TimeUnit2` and that people/code start adapting and replace their usage would be moving forward, which was the gist of the PR...

Sorry for the long-lived PRs :-(

jglick commented on core/src/main/java/hudson/util/TimeUnit2.java:
```diff
@@ -63,7 +67,12 @@
  */
+@Deprecated
+@RestrictedSince("TODO")
+@Restricted(NoExternalUse.class)
```
A restriction is inappropriate here. `@Deprecated` is appropriate and suffices. The reason for the existence of `@Restricted(NoExternalUse.class)` is to allow access from another package within the same source root, similar to `public` in a nonexported package in Java 9.

jglick commented on core/src/main/java/hudson/util/TimeUnit2.java:
```diff
@@ -63,7 +67,12 @@
  *
  * @since 1.5
  * @author Doug Lea
+ * @deprecated use {@link TimeUnit}. (Java 5 did not have all the units required, so TimeUnit2 was introduced because it
```
`@code` or `@link`

batmat commented:
@reviewbybees done

batmat committed:
Add missing @link

batmat commented:
Thanks for the reviews!

daniel-beck commented:
Since `TimeUnit2` is used in plenty of plugins, and newly restricted core API, a changelog entry is definitely needed.

batmat commented:
@daniel-beck 

* `hudson.util.TimeUnit2` deprecated, use `java.util.concurrent.TimeUnit` instead. 

?
